### PR TITLE
Change MAML to Relation Network

### PR DESCRIPTION
**Pull request by ktcshust**
### **User description**
I want change a gradient-based pipeline into a metric-based pipeline with higher performance and accuracy

## Summary by Sourcery

Switch the few-shot learning pipeline from gradient-based MAML to a metric-based Relation Network with a ResNet-18 backbone, stronger augmentations, and a streamlined training loop with cosine annealing LR scheduling

New Features:
- Replace MAML meta-learning implementation with a metric-based Relation Network model using a ResNet-18 encoder and projection head
- Enhance data augmentation pipeline with random vertical flip, stronger rotations, and color jitter
- Introduce a cosine annealing learning rate scheduler for training

Enhancements:
- Increase training epochs to 80, few-shot episodes to 400, and raise embedding and relation dimensions
- Refactor training and evaluation into train_epoch and evaluate functions and consolidate into a RelationNet class
- Switch to AdamW optimizer with weight decay and apply gradient clipping

___

### **PR Type**
Enhancement

___

### **Description**
- Replace gradient-based MAML with Relation Network

- Integrate ResNet-18 embedding and projection head

- Add stronger image augmentations pipeline

- Use AdamW optimizer with CosineAnnealingLR

___

### **Changes diagram**

```mermaid
flowchart LR
  MAML["MAML meta-learning removed"]
  RelNet["Relation Network pipeline"]
  MAML -- "replaced by" --> RelNet
  RelNet --> Emb["ResNet-18 encoder + projection"]
  RelNet --> Aug["Enhanced data augmentations"]
  RelNet --> Opt["AdamW optimizer"]
  RelNet --> Sch["CosineAnnealingLR scheduler"]
```

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Refactor to metric-based Relation Network</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

main.py

<li>Removed MAML implementation<br> <li> Added ResNet-18 embedding class<br> <li> Implemented Relation Network model<br> <li> Updated training with CosineAnnealingLR and AdamW

</details>

  </td>
  <td><a href="https://github.com/ktcshust/dadx/pull/1/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1">+182/-134</a></td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by ktcshust** [comment]
Preparing review...

**Comment by ktcshust** [comment]
## PR Reviewer Guide 🔍

#### (Review updated until commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96)


Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/ktcshust/dadx/pull/1/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R186-R188'><strong>DataLoader Workers</strong></a>

Using num_workers=4 with PathMNIST may fail on some environments (e.g., Windows, limited resources). Consider making it configurable or defaulting to 0/1 to avoid dataloader hangs.
</summary>

```python
train_loader = DataLoader(train_episodes, batch_size=1, shuffle=True, num_workers=4, pin_memory=True)
val_loader = DataLoader(val_episodes, batch_size=1, shuffle=False, num_workers=4, pin_memory=True)
test_loader = DataLoader(test_episodes, batch_size=1, shuffle=False, num_workers=4, pin_memory=True)
```

</details>

<details><summary><a href='https://github.com/ktcshust/dadx/pull/1/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R259-R266'><strong>Scheduler T_max</strong></a>

CosineAnnealingLR with T_max set to epochs steps per epoch once may not reflect per-iteration scheduling. If per-epoch stepping is intended, confirm; otherwise consider T_max=epochs*len(train_loader) and stepping each iteration.
</summary>

```python
optimizer = torch.optim.AdamW(model.parameters(), lr=config.lr, weight_decay=config.weight_decay)
scheduler = CosineAnnealingLR(optimizer, T_max=config.epochs, eta_min=1e-6)

best_val_acc = 0
for epoch in range(config.epochs):
    train_loss, train_acc = train_epoch(model, train_loader, optimizer, config)
    val_loss, val_acc = evaluate(model, val_loader, config)
    scheduler.step()
```

</details>

<details><summary><a href='https://github.com/ktcshust/dadx/pull/1/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R145-R164'><strong>Mixed Episode Sizes</strong></a>

Relation network infers n_way from support_labels; ensure episodes always sample exactly n_way classes and k_shot per class. Add assertions to catch class imbalance or insufficient samples per class at runtime.
</summary>

```python
def forward(self, query_features, support_features, support_labels):
    n_way = len(torch.unique(support_labels))
    k_shot = support_features.size(0) // n_way
    n_query = query_features.size(0)

    # Tính prototype cho mỗi class
    support_features = support_features.view(n_way, k_shot, -1)
    prototypes = support_features.mean(dim=1)

    # Tạo các cặp query-prototype
    query_features = query_features.unsqueeze(1).expand(-1, n_way, -1)
    prototypes = prototypes.unsqueeze(0).expand(n_query, -1, -1)

    # Kết hợp features
    combined = torch.cat([query_features, prototypes], dim=2)
    combined = combined.view(-1, self.config.embedding_dim * 2)

    # Tính relation scores
    relation_scores = self.relation_net(combined)
    return relation_scores.view(n_query, n_way)
```

</details>

</td></tr>
</table>

**Comment by ktcshust** [comment]
## PR Code Suggestions ✨



<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Remove duplicate main call</summary>

___


**There are two identical <code>if __name__ == "__main__":</code> blocks causing <code>main()</code> to run <br>twice; remove the duplicate block to prevent double execution.**

[main.py [282-287]](https://github.com/ktcshust/dadx/pull/1/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R282-R287)

```diff
 if __name__ == "__main__":
     main()
 
-if __name__ == "__main__":
-    main()
-
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Having two `if __name__ == "__main__":` blocks causes `main()` to execute twice, so removing the duplicate prevents unintended double execution.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Fix misplaced print indentation</summary>

___


**The print statement is always executed because it's outdented; indent it inside the <br>if so it only runs when a new best model is saved.**

[main.py [272-275]](https://github.com/ktcshust/dadx/pull/1/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R272-R275)

```diff
 if val_acc > best_val_acc:
     best_val_acc = val_acc
     torch.save(model.state_dict(), 'best_relationnet_resnet.pth')
-print("Saved new best model")
+    print("Saved new best model")
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: The `print("Saved new best model")` is outside the `if` and executes every epoch, moving it inside ensures it runs only when a new best model is saved.


</details></details></td><td align=center>Low

</td></tr><tr><td rowspan=1>General</td>
<td>



<details><summary>Remove unused config instantiation</summary>

___


**The top‐level <code>config = Config()</code> is never used in the module; remove it to avoid <br>confusion and unused instantiation.**

[main.py [32]](https://github.com/ktcshust/dadx/pull/1/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R32-R32)

```diff
-config = Config()
 
+
```
<details><summary>Suggestion importance[1-10]: 3</summary>

__

Why: The top-level `config = Config()` is never used elsewhere and can be removed to reduce confusion and unused code.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

**Comment by ktcshust** [comment]
**[PR Description](https://github.com/ktcshust/dadx/pull/1)** updated to latest commit (https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96)

**Comment by ktcshust** [comment]
**[Persistent review](https://github.com/ktcshust/dadx/pull/1#issuecomment-3067793604)** updated to latest commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96

**Comment by ktcshust** [comment]
Preparing review...

**Comment by ktcshust** [comment]
**[Persistent review](https://github.com/ktcshust/dadx/pull/1#issuecomment-3067793604)** updated to latest commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96

**Comment by ktcshust** [comment]
Preparing review...

**Comment by ktcshust** [comment]
**[Persistent review](https://github.com/ktcshust/dadx/pull/1#issuecomment-3067793604)** updated to latest commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96

**Comment by ktcshust** [comment]
**[Persistent review](https://github.com/ktcshust/dadx/pull/1#issuecomment-3067793604)** updated to latest commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96

**Comment by ktcshust** [comment]
Preparing review...

**Comment by ktcshust** [comment]
Preparing review...

**Comment by ktcshust** [comment]
**[Persistent review](https://github.com/ktcshust/dadx/pull/1#issuecomment-3067793604)** updated to latest commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96

**Comment by ktcshust** [comment]
Preparing review...

**Comment by ktcshust** [comment]
**[Persistent review](https://github.com/ktcshust/dadx/pull/1#issuecomment-3067793604)** updated to latest commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96

**Comment by ktcshust** [comment]
Preparing review...

**Comment by ktcshust** [comment]
**[Persistent review](https://github.com/ktcshust/dadx/pull/1#issuecomment-3067793604)** updated to latest commit https://github.com/ktcshust/dadx/commit/e408ccb1e7980cf60650432e382de04629c2ce96
